### PR TITLE
JOSResampleReader performance and faster quality settings

### PR DESCRIPTION
**Pull request by aras-p**
Backstory: I was looking at what takes up time in Blender during rendering out an animation from a Sequencer. Audio resampling inside Audaspace was "huh feels somewhat costly?" impression.

Currently in Audaspace there's a choice between linear resampler (okay for preview, but not great for final mix), or "extremely high quality preset" for rendering the final mix; with nothing in between. Now, I don't know whether inside say Blender it should be a user preference (like what Audacity has) or a project setting. Based on my findings below, I can't really tell any difference between the current High JOS resampler and the newly added Medium JOS resampler setting, except that the medium one is quite a bit faster.

The filter parameters used:
- High (i.e. current, unchanged): Nz = 136, L = 2304, freq = 0.963904, Kaiser Window B = 16
- Medium: Nz = 42, L = 500, freq = 0.916636, Kaiser Window B = 12
- Low: Nz = 16, L = 128, freq = 0.834068, Kaiser Window B = 10

### Performance

Timings, rendering out 300 frames (frames 1500-1800) of Sprite Fright Edit blender studio data set. Whole rendering process takes about 15 seconds. If using the newly added Medium quality setting, instead of currently existing Audaspace code, the time taken by `JOSResampleReader::read` during rendering:
- Windows (Ryzen 5950X, VS2022): 1.18s -> 0.51s
- Mac (M1 Max, clang 15): 2.68s -> 1.07s

i.e. this change saves about 10% of total rendering time on a Mac.

### Spectrograms

Spectrograms of `Swept_float.wav` from https://src.infinitewave.ca/ in Audacity, using these visualization settings:

jos-spectrograms

jos-resampler

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -101,6 +101,8 @@ set(SRC
 	src/respec/ConverterReader.cpp
 	src/respec/JOSResample.cpp
 	src/respec/JOSResampleReaderCoeff.cpp
```
Comment (neXyon) [inline CMakeLists.txt:103]: If you introduce other qualities, please rename this file as well (adding `High`).

File: include/respec/JOSResampleReader.h [hunk 2]
```diff
@@ -35,20 +35,30 @@ class AUD_API JOSResampleReader : public ResampleReader
...
 
+	static const int m_len_high;
+	static const int m_len_medium;
+	static const int m_len_low;
+	static const int m_L_high;
+	static const int m_L_medium;
+	static const int m_L_low;
+	static const float m_coeff_high[];
+	static const float m_coeff_medium[];
+	static const float m_coeff_low[];
```
Comment (neXyon) [inline include/respec/JOSResampleReader.h:46]: Please add documentation for these.

File: include/respec/JOSResampleReader.h [hunk 3]
```diff
@@ -111,13 +121,23 @@ class AUD_API JOSResampleReader : public ResampleReader
 	void AUD_LOCAL resample_mono(double target_factor, int length, sample_t* buffer);
 	void AUD_LOCAL resample_stereo(double target_factor, int length, sample_t* buffer);
 
+	template <typename T>
+	void AUD_LOCAL do_resample(double target_factor, int length, sample_t* buffer);
```
Comment (neXyon) [inline include/respec/JOSResampleReader.h:125]: I recommend renaming this to `resample` and the currently same called function to `resample_generic`.

File: include/respec/JOSResampleReader.h [hunk 4]
```diff
@@ -111,13 +121,23 @@ class AUD_API JOSResampleReader : public ResampleReader
...
+
 public:
+	enum Quality
```
Comment (neXyon) [inline include/respec/JOSResampleReader.h:128]: Could use `enum class` (yeah, I know, isn't used in the rest of the library, but I have improved my C++ skills in the meantime :D) and please add documentation.

File: src/respec/JOSResampleReader.cpp [hunk 5]
```diff
@@ -20,9 +20,27 @@
...
+static inline int lrint_impl(double x)
+{
+	return _mm_cvtsd_si32(_mm_load_sd(&x));
+}
+#else
+static inline int lrint_impl(double x)
+{
+	return lrint(x);
+}
+#endif
```
Comment (neXyon) [inline src/respec/JOSResampleReader.cpp:38]: I don't see the need for defining `HAVE_SSE2` since you can do everything in the first `#if` already?

**Review by neXyon — CHANGES_REQUESTED** [review]
Thanks for this very nice patch! My inline comments are all minor, however I would ask for two things before approving:

1. What speed improvement do you get from the SSE2 optimization? I.e. with and without it using the high quality filter.
2. I am not a fan of copying the filters from libsamplerate and would rather have filters designed similarly to the one I created for audaspace. Originally, I designed them with octave (open source implementation of matlab), but I have ported the filter design script to python which gives a similar output and pushed it here: https://github.com/audaspace/audaspace/blob/master/src/respec/filter_design.py Maybe you can check that out and try to get a kaiser windowed sinc filter for medium and low quality as well?

**Comment by aras-p** [comment]
> 1. What speed improvement do you get from the SSE2 optimization

You mean from SSE2 lrint? It's in the message of that commit. Resampling 3min30sec 44.1khz stereo wav file into 48khz one, on Windows / Ryzen 5950X: 7.92s -> 7.33s.

I guess rewriting whole function to be SIMD might make it yet faster, but that would be way more invasive change.

> 2. I am not a fan of copying the filters from libsamplerate and would rather have filters designed similarly to the one I created for audaspace

Makes sense, indeed I mostly have no idea what I'm doing :) How would I design medium/low quality filter using your script? It has three parameters, that's a large parameter space to explore for someone who has no idea about them.

**Comment by neXyon** [comment]
> > 1. What speed improvement do you get from the SSE2 optimization
> 
> You mean from SSE2 lrint? It's in the message of that commit. Resampling 3min30sec 44.1khz stereo wav file into 48khz one, on Windows / Ryzen 5950X: 7.92s -> 7.33s.
> 
> I guess rewriting whole function to be SIMD might make it yet faster, but that would be way more invasive change.

Ah, so about a 7.5% decrease in time. That may not sound much, but it's significant.

> > 2. I am not a fan of copying the filters from libsamplerate and would rather have filters designed similarly to the one I created for audaspace
> 
> Makes sense, indeed I mostly have no idea what I'm doing :) How would I design medium/low quality filter using your script? It has three parameters, that's a large parameter space to explore for someone who has no idea about them.

Well, you can see the parameters that libsamplerate used. `L` is the `coeffs` variable you had to give the code anyway, so that should have a similar value. The number of values in the array then mostly depends on `Nz * L`, so you get an approximate `Nz` by dividing the length of the array by `L`. Since the length of the array also determines the speed of your computations, a lower number will give you a faster results, but the quality will of course be poorer. And then for a given `L` and `Nz` you want to set the window parameter `B` in a sensible way. From [wikipedia](https://en.wikipedia.org/wiki/Kaiser_window):

> ... is a non-negative real number that determines the shape of the window. In the frequency domain, it determines the trade-off between main-lobe width and side lobe level, which is a central decision in window design.

The ideal filter would be 1 (= 0 dB) for 0 to 1 Hz and then drop to 0 (- inf dB), but that would require an infinitely long filter, so that's not achievable. So instead you want to stay at 0 dB as long as possible and then drop off as sharply as possible without ripples or a spike that goes above 0 dB and then stay as low as possible at below -1XX dB usually, that's often called the signal-to-noise ratio (SNR). `B` lets you trade off how long it stays at 0 dB/how quickly it goes down vs. how low it gets (and stays).

To get approximately something that libsamplerate uses, try these to start with:

- medium: `L = 500`, `Nz = 42`, `B = 12`
- low: `L = 128`, `Nz = 16`, `B = 10`

From here it's an art... Basically you you want to make sure that your passband is as nice as possible (again: flat at 0 as long as possible, no ripples and then going down as quickly as possible) given a specific length of the array and the noise floor dB you want. That's what the plots are trying to show you, since "Passband" and "Transition" are just zooming in on the relevant parts of the "Magnitude" plot. You can ignore the "Response" and "Phase" plots. The latter should be fine if you only vary the 3 parameters and don't try other filter design methods/windows/etc.

**Comment by aras-p** [comment]
Thanks for the info! I'll try those... next year :P

**Comment by neXyon** [comment]
Thanks! I see you were happy with my suggested values already ;)

**Comment by aras-p** [comment]
> I see you were happy with my suggested values already ;)

Yeah I played around with several neighboring parameter values but the filter graphs did not look better.

**Pull request merged**